CH4 RMA Patches

shawnccx opened the pull request:
This PR contains a few CH4 RMA patches we added to our internal repository recently, including:

1. Fix label in OFI `win_free`.
2. Eliminate OFI netmod-specific RMA synchronization.
3. Move OFI map functions to CH4 as generic ones, and replace `MPL_HASH` functions (4 commits here).
4. Dynamic RMA window target allocation scheme.

raffenet commented:
test:jenkins/ch4/all

raffenet commented:
This looks to be causing a regression in RMA tests for ch4/ucx. E.g. https://jenkins.mpich.org/job/mpich-review-ch4-ucx/537/compiler=gnu,jenkins_configure=debug,label=ib64/testReport/(root)/summary_junit_xml/373_____rma_putpscw1_4/

shawnccx commented:
I have located this issue, we forgot to consider the case a target lock object is NULL during RMA in AM level (which is fine in OFI level), fixing it now.

min-si commented:
Thanks. I can review on the commits tomorrow.

shawnccx commented:
@raffenet can u help me trigger the tests again? just updated my commits.

min-si commented:
test:jenkins/ch4/all

min-si commented:
test:jenkins/ch4/all

min-si commented:
test:jenkins/ch4/all

shawnccx commented:
Updated a more efficient way (by using `MPL_HASH_ITER`) to clean hash table in newest commit.

raffenet commented:
test:jenkins/ch4/all

raffenet commented:
Not sure why Min's comments didn't trigger the test. Need to look into that...

hajimefu commented:
@pkcoff You might also want to try this out because this makes some changes to RMA synchronization protocol.

min-si commented:
@raffenet @shawnccx : Looks all jenkins tests passed. I will review the patch from tomorrow. Sorry that I had to deal with other things till tonight thus has to delay review for this.

shawnccx commented:
I just rebased on pmodels/master to solve recent merge conflicts.

shawnccx commented:
@raffenet can you trigger tests again to check if there is error after this patch is rebased to newest master branch.

raffenet commented:
test:jenkins/ch4/most

min-si commented on src/mpid/ch4/src/ch4r_win.h:
```diff
@@ -515,7 +500,8 @@ static inline int MPIDI_CH4R_mpi_win_lock(int lock_type, int rank, int assert, M
 
-    MPIDI_CH4U_win_target_t *target_ptr = &MPIDI_CH4U_WIN(win, targets)[rank];
+    MPIDI_CH4U_win_target_t *target_ptr = MPIDI_CH4U_win_target_add(win, rank);
+
```
Thanks for making this optimization. Let me try to summarize the design of target object allocation. Please correct me if my understanding is wrong:
1. Always create a target at `lock`, and free it at `unlock`.
2. For `lock_all`, `lock`, or `start-complete` epoch, create a target object at operation issuing time if such object does not exist.
3. Always clean all target objects in window at `unlock_all`, `win_complete`.
4. In `fence`, clean all target objects in window if `assert` does not include `MPI_MODE_NOPRECEDE`.

Here are my questions:
1. This design is very similar to the memory optimizing policy in CH3. But I am concerning performance drop if we exactly follow the current policy in CH4 which is much more latency sensitive than CH3. To be specific, for a program that does frequent lock-unlock on same group of targets, creation-freeing target at every epoch would obviously increase latency. Can we reuse the allocated target object, and just release all at `win_free` ? Or, at least we should allow user to pass `info hint` to us whether the program prefers a `memory efficient` mode or a `performance efficient` mode.
2. Why clean all target objects in `fence without MPI_MODE_NOPRECEDE` ? Any existing target object should be freed by the previous epoch closing call (if the user program is correct), right ?

shawnccx replied:
@min-si , your understanding is correct.
For your questions,
1. It is a trade-off between memory cost and lock/unlock cost.
I talked with @hajimefu , yes, one option is reusing existing target obj and freeing all at `win_free`. But in this case, an application requiring locking for multiple targets (just an extreme opposite example to your case) would introduce high memory cost.

My suggestion is using current implementation, and create an enhancement ticket about `info hint` (I can fix that when my current workload is done, should not take much time).

Anyway, it is up to you.

2. For `win_fence`, my understanding is that `MPI_MODE_NOPRECEDE` assertion means `no RMA completion` (from book `Using advanced MPI`). So there is no need to clean all existing `targets` as all ops are supposed not to be done.

min-si replied:
1. The `info hint` plan sounds good to me. Thanks !

2. In `fence` epoch, your implementation does not add per-target object at operation issuing time,  so you should not need the cleanup at a `fence` call. The only case that would cause existence of per-target objects at `fence` epoch is incorrect user program, e.g., user calls `start`->'put (per-target object is created here)'->'fence'. But the `MPIDI_CH4U_FENCE_EPOCH_CHECK` sync check should catch such case, and throw error if user does not disable `error-checking` at MPICH configure time. If user disables `error-checking`, an incorrect user program causes undefined behavior, thus you do not need to free those target objects either.

shawnccx commented:
Seems lots of failures happens after rebasing. I'll look into it.

shawnccx commented on src/mpid/ch4/netmod/ofi/ofi_win.h:
```diff
@@ -269,18 +269,14 @@ static inline int MPIDI_OFI_win_progress_fence(MPIR_Win * win)
 static inline int MPIDI_NM_mpi_win_set_info(MPIR_Win * win, MPIR_Info * info)
 {
     int mpi_errno;
+
```
will fix these in next commit.  Existing code is not unified for with/without empty line. So I just add one empty line for all here.

min-si replied:
I see. But please move all format cleanup changes in a separate commit.

shawnccx commented on src/mpid/ch4/netmod/ofi/util.c:
```diff
@@ -418,65 +261,27 @@ int MPIDI_OFI_control_handler(int handler_id, void *am_hdr,
 {
-    int senderrank;
     int mpi_errno = MPI_SUCCESS;
     void *buf = am_hdr;
```
i can put removing `void *buf = am_hdr;` into the commit of cleaning unused  util.c functions

shawnccx commented:
located the failure, will update with address comments tomorrow.

shawnccx commented on src/mpid/ch4/netmod/ofi/ofi_win.h:
```diff
@@ -376,63 +303,21 @@ static inline int MPIDI_OFI_fill_ranks_in_win_grp(MPIR_Win * win_ptr, MPIR_Group
-
-    MPIR_Group_release(MPIDI_CH4U_WIN(win, sync).sc.group);
-    MPIDI_CH4U_WIN(win, sync).sc.group = NULL;
+    mpi_errno = MPIDI_CH4R_mpi_win_complete(win);
```
I will change all related code to this style. It would eliminate one extra line of `MPIDI_CH4R_win_func`.

shawnccx commented on src/mpid/ch4/src/ch4r_win.h:
```diff
@@ -730,7 +715,7 @@ static inline int MPIDI_CH4R_win_finalize(MPIR_Win ** win_ptr)
     } while (all_completed != 1);
 
-    MPL_free(MPIDI_CH4U_WIN(win, targets));
+    MPIDI_CH4U_win_hash_clear(win);
```
In order to make sure all targets are released, I agree with releasing existing targets are `win_finalize`.

shawnccx commented:
@min-si I've addressed your comments in recent commits, including:

1. remove changes of empty lines at this time.
2. reduce code duplication for ofi level window synchronization functions calling `MPIDI_CH4R_win_func`.
3. free target objs in `win_finalize`, and remove unnecessary free in `win_fence`.

@raffenet Please help me trigger tests, thanks.

min-si commented:
test:jenkins/ch4/all

raffenet commented:
test:jenkins/ch4/all

shawnccx commented:
Quite confused by jenkins builds result here. `mpich-review-ch4-armci-mpi` failed in building MPICH while the other three succeeded in building MPICH. @raffenet Do you have any comments for `armci` build failure?

shawnccx commented:
Also, this patch doesn't touch anything in `comm` or `threads` parts.
In ch4-ofi build, it looks like ` ./threads/comm/idup_comm_gen 4` is an intermittent failure. I tested `master` branch on local machine, it failed once in 10 runs.

raffenet commented:
@shawnccx I'm not sure what happened there. Compile error doesn't really make any sense. I think I need to look at the job and make sure it's getting the right branches...

raffenet commented:
Made a tweak to the armci job configuration. Let's try again.
test:jenkins/ch4/armci-mpi

raffenet commented:
One more tweak to tell Jenkins how to access PR branches.
test:jenkins/ch4/armci-mpi

raffenet commented:
@min-si @shawnccx OK, Jenkins is testing the right branch now. Turns out the armci tests have been broken with ch4/ofi since September 26th, though. https://jenkins.mpich.org/view/nightly-master-ch4/job/mpich-master-ch4-armci-mpi/164/. I think it's safe to ignore those test results this time.

min-si commented:
It fails because `contiguous-bench` hangs. I can do a quick check on this.

hajimefu commented:
It was introduced by one of our patches (671dce72). We are currently looking at it.

min-si commented:
Ah, OK. I will make a pass on this PR again later today or tomorrow morning.

min-si commented:
@shawnccx : Just noticed some of your commits are with `Author: NIS <chongxiao.cao@intel.com>`. Is this a typo ?

shawnccx commented:
@min-si  thanks for pointing it out. fixed.

min-si commented:
test:jenkins/ch4/all

minsii commented:
test:jenkins/ch4/all

minsii commented:
test:jenkins/ch4/armci-mpi

minsii commented:
test:jenkins/ch4/armci-mpi

raffenet commented:
test:jenkins/ch4/all

raffenet commented:
I'm still in the process of getting Jenkins results in a good state. These results look fine. No unexpected failures.

minsii commented:
@raffenet : thanks. I will rebase this PR.
